otelcol/command_print.go

```diff
@@ -4,55 +4,205 @@
+
+	modeHelp := "Operating mode: redacted (default), unredacted"
+	cmd.Flags().StringVar(&mode, "mode", "redacted", modeHelp)
+
+	validateHelp := "Validation mode: true (default), false"
+	cmd.Flags().BoolVar(&validate, "validate", true, validateHelp)
+
 	cmd.Flags().AddGoFlagSet(flagSet)
 	return cmd
 }
+
+type printContext struct {
+	cmd          *cobra.Command
+	stdout       io.Writer
+	stderr       io.Writer
+	set          CollectorSettings
+	outputFormat string
+	validate     bool
+}
+
+func (pctx *printContext) configPrintSubCommand(flagSet *flag.FlagSet, mode string) error {
+	if !printCommandFeatureFlag.IsEnabled() {
+		return errors.New("print-config is currently experimental, use the otelcol.printInitialConfig feature gate to enable this command")
+	}
+	err := updateSettingsUsingFlags(&pctx.set, flagSet)
+	if err != nil {
+		return err
+	}
+
+	if !pctx.validate {
+		fmt.Fprintf(pctx.stderr, "Warning: redacted mode hides sensitive fields but does not validate. Use with caution.\n")
+	}
+	switch strings.ToLower(mode) {
+	case "redacted":
+		return pctx.printRedactedConfig()
+	case "unredacted":
+		fmt.Fprintf(pctx.stderr, "Warning: unredacted mode shows the complete configuration. Use with caution.\n")
+		return pctx.printUnredactedConfig()
+	default:
+		return fmt.Errorf("invalid mode %q: modes are: redacted, unredacted", mode)
+	}
+}
+
+// printConfigData formats and prints configuration data in yaml or json format.
+func (pctx *printContext) printConfigData(data map[string]any) error {
+	format := pctx.outputFormat
+	if format == "" {
+		format = "yaml"
+	}
+
+	if strings.EqualFold(format, "json") {
+		// Note that JSON does
+		fmt.Fprintf(pctx.stderr, "Warning: JSON output format is unstable. Use with caution.\n")
+		encoder := json.NewEncoder(pctx.stdout)
+		encoder.SetIndent("", "  ")
+		return encoder.Encode(data)
+	}
+
+	if strings.EqualFold(format, "yaml") {
+		b, err := yaml.Marshal(data)
+		if err != nil {
+			return err
+		}
+		fmt.Fprintf(pctx.stdout, "%s\n", b)
+		return nil
+	}
+	return fmt.Errorf("unrecognized print format: %s", format)
+}
+
+func (pctx *printContext) printConfiguration() (any, error) {
+	var factories Factories
+	if pctx.set.Factories != nil {
+		var err error
+		factories, err = pctx.set.Factories()
+		if err != nil {
+			return nil, fmt.Errorf("failed to get factories: %w", err)
+		}
+	}
+
+	configProvider, err := NewConfigProvider(pctx.set.ConfigProviderSettings)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create config provider: %w", err)
+	}
+
+	cfg, err := configProvider.Get(pctx.cmd.Context(), factories)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get config: %w", err)
+	}
+	return cfg, nil
+}
+
+// printUnredactedConfig prints resolved configuration before interpreting
+// with the intended types for each component, thus it shows the full
+// configuration without considering configuopaque. Use with caution.
+func (pctx *printContext) printUnredactedConfig() error {
```

Comment:
Can this be a CVE? I am thinking of a scenario where via the otelcol someone gets access to some credentials (VoltDB, etc.)?

I have not spent too much time thinking of all possible scenarios but I am worried a bit to offer this.